### PR TITLE
expression: support tidb encode key function

### LINKED ISSUE #51683: Support querying MVCC information through builtin function
## Enhancement

MVCC information is useful for debugging data inconsistenct issue. For now, we have HTTP API `mvcc/key/...` to query MVCC. However, it is not always convenient to construct a HTTP query. For example, when there are complex data types in index columns or clustered primary key columns like `timestamp`, `datetime` or even `binary`, query parameters cannot represent these values.

Maybe we can provide a builtin function like `tidb_mvcc_info()` achieve this purpose.

Note: the priviege should be considered to make sure there are no security issues.

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #51683

Problem Summary: See #51683.

### What changed and how does it work?

This PR supports three builtin functions, namely `tidb_mvcc_info()`, `tidb_encode_record_key()` and `tidb_encode_index_key()`:

- `tidb_mvcc_info(key)` returns MVCC information of a given key.
- `tidb_encode_record_key(dbName, tableName, handle/pk columns...)` returns a row record key. The table must exist.
- `tidb_encode_index_key(dbName, tableName, indexName, idx columns..., handle/pk columns...)` returns a index record key. The index must exist.

Please also note that *tableName* can be one of formats as below:
- `t`: refer to non-partitioned table `t`.
- `t(p1)`: refer to partition table `t` and the partition `p1`.

```sql
mysql> create table t (a int, b int, c timestamp, primary key (a) clustered, index idx_c(c));
Query OK, 0 rows affected (0.14 sec)

mysql> insert into t values (1, 1, now());
Query OK, 1 row affected (0.00 sec)

mysql> table t;
+---+------+---------------------+
| a | b    | c                   |
+---+------+---------------------+
| 1 |    1 | 2024-03-12 11:14:35 |
+---+------+---------------------+
1 row in set (0.00 sec)

mysql> select tidb_encode_record_key('test', 't', 1);
+----------------------------------------+
| tidb_encode_record_key('test', 't', 1) |
+----------------------------------------+
| 7480000000000000705f728000000000000001 |
+----------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_mvcc_info('7480000000000000705f728000000000000001');
+-----------------------------------------------------------------------------------------------------------------------------------+
| tidb_mvcc_info('7480000000000000705f728000000000000001')                                                                          |
+-----------------------------------------------------------------------------------------------------------------------------------+
| {"info":{"writes":[{"start_ts":448322148825038849,"commit_ts":448322148825038850,"short_value":"gAACAAAAAgMBAAkAAQAAAKMz2LIZ"}]}} |
+-----------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_encode_index_key('test', 't', 'idx_c', '2024-03-12 11:14:35', 1);
+----------------------------------------------------------------------------+
| tidb_encode_index_key('test', 't', 'idx_c', '2024-03-12 11:14:35', 1)      |
+----------------------------------------------------------------------------+
| 7480000000000000705f6980000000000000010419b2d833a3000000038000000000000001 |
+----------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_decode_key(tidb_encode_index_key('test', 't', 'idx_c', '2024-03-12 11:14:35', 1));
+----------------------------------------------------------------------------------------+
| tidb_decode_key(tidb_encode_index_key('test', 't', 'idx_c', '2024-03-12 11:14:35', 1)) |
+----------------------------------------------------------------------------------------+
| {"index_id":1,"index_vals":{"c":"2024-03-12 11:14:35"},"table_id":112}                 |
+----------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_mvcc_info(tidb_encode_index_key('test', 't', 'idx_c', '2024-03-12 11:14:35', 1));
+-----------------------------------------------------------------------------------------------------------+
| tidb_mvcc_info(tidb_encode_index_key('test', 't', 'idx_c', '2024-03-12 11:14:35', 1))                     |
+-----------------------------------------------------------------------------------------------------------+
| {"info":{"writes":[{"start_ts":448322148825038849,"commit_ts":448322148825038850,"short_value":"MA=="}]}} |
+-----------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by ekexium** [comment]
I love this
We need this
Please do continue this

**Comment by D3Hunter** [comment]
i hope those functions would works in normal query like :
```
select tidb_handle(), tidb_idx_key(index-name) from test.t where id = 1
select tidb_mvcc(tidb_handle()) from test.t where id = 1
```

**Comment by tangenta** [comment]
@D3Hunter If we put this in normal query, I think we cannot get the MVCC of deleted rows/indexes?

**Comment by kennytm** [comment]
@D3Hunter I don't think a function should need or be able to know the current plan is a SELECT operating on exactly one physical table.

**Review by kennytm — CHANGES_REQUESTED** [review]
This must not be merged before the privilege check bypass is fixed. 

See https://github.com/pingcap/tidb/issues/51683#issuecomment-2168137763 for a PoC.

**Comment by D3Hunter** [comment]
> i hope those functions would works in normal query like :
> 
> ```
> select tidb_handle(), tidb_idx_key(index-name) from test.t where id = 1
> select tidb_mvcc(tidb_handle()) from test.t where id = 1
> ```

ok, for downvoters, get from query is much simplier, i don't like to input so many parameters to just get the key of it, and it's error-prone to do that as we have to remember the definition of the pk/index

> If we put this in normal query, I think we cannot get the MVCC of deleted rows/indexes?

can use `as of timestamp`?

> I don't think a function should need or be able to know the current plan is a SELECT operating on exactly one physical table.

don't get your point, where is `a function should need or be able to know the current plan is a SELECT operating on exactly one physical table.` rule came from?

**Comment by kennytm** [comment]
@D3Hunter what is the result of `tidb_handle()` in the following?

```sql
select tidb_handle() from a, b;
select tidb_handle() from dual;
select tidb_handle() from (select * from t) _;

create view v as select 1;
select tidb_handle() from v;
```

**Comment by D3Hunter** [comment]
> @D3Hunter what is the result of `tidb_handle()` in the following?
> 
> ```sql
> select tidb_handle() from a, b;
> select tidb_handle() from dual;
> select tidb_handle() from (select * from t) _;
> 
> create view v as select 1;
> select tidb_handle() from v;
> ```

if we use the pattern, just check and report error in this case

**Comment by kennytm** [comment]
> if we use the pattern, just check and report error in this case

No SQL flavor (MySQL or Postgres or T-SQL etc) would determine the validity of a function by the chosen query plan.

On TiDB there is a single exception of `tidb_row_checksum()` (#43479) which the PR description claimed this **is a quick & dirty implementation**, and I absolutely oppose making this the future convention. A pseudo-column that attaches to the table object i.e. `t._tidb_rowid` makes much more sense than the plan-dependent `tidb_row_checksum()` model (even if both are hacks).

**Comment by D3Hunter** [comment]
> No SQL flavor (MySQL or Postgres or T-SQL etc) would determine the validity of a function by the chosen query plan.

unfair to compare with them, they don't have the requirements to access underlying encoded DATA(KV) too

> absolutely oppose

please give insights, now just your standpoint

**Comment by kennytm** [comment]
@D3Hunter

> unfair to compare with them, they don't have the requirements to access underlying encoded DATA(KV) too

T-SQL has informational functions like [OBJECT_ID](https://learn.microsoft.com/en-us/sql/t-sql/functions/object-id-transact-sql?view=sql-server-ver16) to obtain the low-level information and using similar semantic as the current proposal --- identify the table via a string

```sql
select object_id('db.schema.tbl');
```

Similar for [PostgreSQL](https://www.postgresql.org/docs/current/functions-info.html) which the functions either return something that can be uniquely identified (`current_schema()`, `pg_current_xact_id()`, ...) or requires user to supply the identity (` col_description(table_oid, col_number), ...`). 

I can't find any function that has implicit assumption about which table the SELECT is working on.

> please give insights, now just your standpoint

Keep the current syntax unchanged `tidb_encode_record_key('schema', 'tbl', pk_val)` and `tidb_encode_index_key('schema', 'tbl', 'idx', idx_vals, pk_val)`.

(One could introduce a pseudo-column `t._tidb_record_key` but it doesn't work for index keys nicely.)

**Comment by kennytm** [comment]
I think OracleDB's [Pseudocolumns](https://docs.oracle.com/en/database/oracle/oracle-database/23/sqlrf/Pseudocolumns.html) are pretty close in spirit. But I don't think there's much mainstream DB products that exposes low-level row encoding via SQL.

Another alternative is a table function like

```sql
select tbl.a, tbl.c, mvcc.record, mvcc.index_idx_c 
from tbl 
join tidb_mvcc_info(tbl) mvcc on true 
where tbl.a = 1;
```

but probably don't bother when we don't even support the MySQL-compatible `JSON_TABLE()` table function :shrug: 

Also both pseudo-column and table function can't give us MVCC info for ***deleted*** rows that are not yet GC'ed which sometimes we may want to access.

BTW shouldn't these syntax questions be better commented on #51683 rather than this PR?

**Comment by tangenta** [comment]
@kennytm I added the privilege check in [this commit](https://github.com/pingcap/tidb/pull/51678/commits/03a9dc0df6b82269b1d1d5766a632cafccb8c271), PTAL

File: pkg/planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -2640,6 +2641,185 @@ func hasCurrentDatetimeDefault(col *model.ColumnInfo) bool {
...
+	if err != nil || isNull {
+		return nil, isNull, err
+	}
+	tblName, isNull, err := args[1].EvalString(ctx, row)
+	if err != nil || isNull {
+		return nil, isNull, err
+	}
+	is := isVer.(infoschema.InfoSchema)
+	if is == nil {
+		return nil, false, errors.New("missing information schema")
```
Comment (kennytm) [inline pkg/planner/core/expression_rewriter.go:2660]:
```suggestion
	is, ok := isVer.(infoschema.InfoSchema)
	if !ok {
		return nil, false, errors.New("missing information schema")
```

the cast will crash if `isVer` is actually nil 

```
panic: interface conversion: interface is nil, not infoschema.InfoSchema
```

File: pkg/planner/core/expression_rewriter.go [hunk 2]
```diff
@@ -2640,6 +2641,185 @@ func hasCurrentDatetimeDefault(col *model.ColumnInfo) bool {
...
+	}
+	tbl, err := is.TableByName(context.Background(), model.NewCIStr(dbName), model.NewCIStr(tblName))
+	if err != nil {
+		return nil, false, err
+	}
+	tblInfo := tbl.Meta()
```
Comment (kennytm) [inline pkg/planner/core/expression_rewriter.go:2779]: is `tbl` actually needed here, or is `TableInfoByName` sufficient (TableInfoByName requires only MetaOnlyInfoSchema).

**Review by kennytm — APPROVED** [review]

File: pkg/planner/core/expression_rewriter.go (hunk 2 above)
Comment (tangenta) [reply pkg/planner/core/expression_rewriter.go]: Yes, we need `table.Table` to choose the correct partition. Previous implementation does not consider partition. This bug is fixed in https://github.com/pingcap/tidb/pull/51678/commits/4e07b96d212ad91e4ca8e1d272c9147a72c8b18b.

File: pkg/planner/core/expression_rewriter.go (hunk 1 above)
Comment (tangenta) [reply pkg/planner/core/expression_rewriter.go]: It cannot be nil, I removed the nil-check.

**Comment by D3Hunter** [comment]
please ask PM for approval for new feature

**Comment by D3Hunter** [comment]
/hold

unhold it after PM approve

**Review by Frank945946 — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Review by D3Hunter — COMMENTED** [review]
~add some test for user privileges?~

**Review by XuHuaiyu — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**